lintcheck/src/json.rs

```diff
@@ -130,9 +143,65 @@ pub(crate) fn diff(old_path: &Path, new_path: &Path) {
     println!();
     println!();
 
-    print_warnings("Added", &added);
-    print_warnings("Removed", &removed);
-    print_changed_diff(&changed);
+    print_warnings("Added", added);
+    print_warnings("Removed", removed);
+    print_changed_diff(changed);
+}
+
+/// This function limits the number of lints in the collection if needed.
+///
+/// It's honestly amazing that a function like this is needed. But some restriction
+/// lints (Looking at you `implicit_return` and `if_then_some_else_none`) trigger a lot
+/// like 60'000+ times in our CI.
+///
+/// Each lint in the list will be limited to 200 messages
+fn limit_by_lint<T, F>(list: &mut Vec<T>, lint_key: F) -> Option<String>
+where
+    F: Fn(&T) -> &str,
+{
+    const PER_LINT_LIMIT: usize = 200;
```

Comment:
A different default might be better. I considered 250 for a while.

For now, I landed on 200 as a big change to a restriction lint might generate X warnings in all sections and I wanted the sum of all changes (so 600 warnings) to be less that GH's limit. 

I also considered lowering it just 100, as I'm guessing there are diminishing returns. I don't think I would skim through more than 50 changes. Any thoughs on what a good limit would be?

Reply:
200 per lint might be a bit much if there several lints changing

We could do something dynamic like the per limit being `200 / number of lints` so if there were 10 different lints being added you'd see up to 20 of each

Or, there's the option of saying this is pretty rare so just truncate the vecs to whatever number

Reply:
I think it's very rare unless bigger refactorings are done to restriction lints. I calculate each message to have less than 1k bytes, so we can display around 1000 warnings in the summery. So 20 are very few IMO.

What do you think about max 200 and if we have more than 5 changed lints it's `1000/number of lints`?

That paired with a pair lint section as described below should still make the logs readable.

<sub>So much for me thinking that this was the last lintcheck PR for some time :P</sub>

Reply:
https://github.com/xFrednet/rust-clippy/actions/runs/10028799118?pr=4 displays 1200ish and comes in at 869KB, considering most of those diagnostics are short it's already cutting it close

I don't mind how the limit is distributed as long as we fit in some reasonable total number, 600 is conservative with respect to the file size limit but human readability wise I think it's plenty

Reply:
I've reworked a good part of the output now. Once everything is ready, I'll push an update.

The truncation is now dynamic and defaults to 300 warnings per lint. This results in 100 warnings per section.

We should still keep the step of first writing to a file and then the summary in case the truncated output exceeds the limit somehow.